app/services/share_variant_builder.rb

```diff
@@ -36,60 +38,79 @@ def variant_initials(variant_type)
+    @button = Share::Button.find_by(sp_type: @variant_type, page_id: @page.id)
+    update_and_return
+  end
 
-    button = Share::Button.find_by(sp_type: @variant_type, page_id: @page.id)
-    sp_button = ShareProgress::Button.new(share_progress_button_params(variant, button))
+  def create
+    @variant = variant_class.new(@params)
+    @variant.page = @page
+    return @variant unless @variant.valid?
+    @button = Share::Button.find_or_initialize_by(sp_type: @variant_type, page_id: @page.id)
+    @variant.button = @button
+    update_and_return
+  end
 
-    if sp_button.save
-      variant.save
-    else
-      add_sp_errors_to_variant(sp_button, variant)
+  def destroy
+    @variant = variant_class.find(@id)
+    @button = Share::Button.find_by(sp_type: @variant_type, page_id: @page.id)
+
+    if @variant.share_progress?
+      sp_button = ShareProgress::Button.new(share_progress_button_params(@variant, @button))
+      sp_variant = sp_variant_class.new(id: @variant.sp_id, button: sp_button)
+      unless sp_variant.destroy
+        add_sp_errors_to_variant(sp_variant, @variant)
+        return @variant
+      end
     end
-    variant
+    @variant.destroy
+    @variant
   end
 
-  def create
-    variant = variant_class.new(@params)
-    variant.page = @page
-    return variant unless variant.valid?
-    button = Share::Button.find_or_initialize_by(sp_type: @variant_type, page_id: @page.id)
-    variant.button = button
-    sp_button = ShareProgress::Button.new(share_progress_button_params(variant, button))
+  private
 
-    if sp_button.save
-      button.update(sp_id: sp_button.id, sp_button_html: sp_button.share_button_html, url: sp_button.page_url)
-      variant.update(sp_id: sp_button.variants[@variant_type].last[:id])
+  def update_and_return
+    if @variant.share_progress?
+      update_sp_resources
     else
-      add_sp_errors_to_variant(sp_button, variant)
+      update_button
+      @variant.save
     end
-    variant
+    @variant
   end
 
-  def destroy
-    variant = variant_class.find(@id)
-    button = Share::Button.find_by(sp_type: @variant_type, page_id: @page.id)
-    sp_button = ShareProgress::Button.new(share_progress_button_params(variant, button))
-    sp_variant = sp_variant_class.new(id: variant.sp_id, button: sp_button)
-    if sp_variant.destroy
-      variant.destroy
-    else
-      add_sp_errors_to_variant(sp_variant, variant)
+  def update_button
+    # Update with other options when we will support other non-sp share types.
+    # Note that the share button html MUST have a css class - otherwise it'll be missed by Shares.get_all
+    # Add a class that indicates the variant type, and in _share.liquid, add a class that identifies the variant/button.
+    case @variant_type
+    when :whatsapp
+      @button.update(
+        sp_button_html: "<a class=\"whatsapp_large\" href=\"whatsapp://send?text=#{CGI.escape(@variant.text)}\"></a>",
+        url: @url
```

Comment:
WhatsApp actually recommends `https://api.whatsapp.com/send?text={{urlencodedtext}}` (as opposed to `whatsapp://`) since that also works on the desktop and is future proof. https://faq.whatsapp.com/en/general/26000030

Reply:
Omg thank you. That explains why after having changed that, it suddenly stopped working on desktop. I thought it was just Ubuntu. :woman_shrugging: